extensions/resource-deployment/src/services/tools/toolBase.ts

```diff
@@ -19,47 +20,170 @@ export abstract class ToolBase implements ITool {
+			this.logToOutputChannel(`   >${command}`);
+			this.logToOutputChannel(`   Could not find 'Location' in the output:`);
+			this.logToOutputChannel(pip3ShowOutput, `   output:`);
+			return '';
+		} else {
+			return retValue;
 		}
-		this._isInstalled = false;
-		this._statusDescription = undefined;
-		this._version = undefined;
-		this._versionOutput = undefined;
-		return this._platformService.runCommand(this.versionCommand).then((stdout) => {
-			this._versionOutput = stdout;
-			this._version = this.getVersionFromOutput(stdout);
-			if (this._version) {
-				this._isInstalled = true;
-			} else {
-				throw localize('deployCluster.InvalidToolVersionOutput', "Invalid output received.");
+	}
+
+	public get outputChannelName(): string {
+		return this._platformService.outputChannelName();
+	}
+
+	public showOutputChannel(preserveFocus?: boolean | undefined): void {
+		this._platformService.showOutputChannel(preserveFocus);
+	}
+
+	public async install(updateDisplayTableData: (tool: ITool) => void, thisObject: any): Promise<void> {
+		try {
+			this._status = ToolStatus.Installing;
+			updateDisplayTableData.call(thisObject, this);
+			await this.installCore();
+			await this.checkAndUpdateVersion();
+			updateDisplayTableData.call(thisObject, this);
+		} catch (error) {
+			const errorMessage = this._platformService.getErrorMessage(error);
+			this._statusDescription = localize('deployCluster.InstallError', "Error installing tool '{0}'.{1}Error: {2}{1}See output channel '{3}' for more details", this.displayName, EOL, errorMessage, this.outputChannelName);
+			this._status = ToolStatus.Error;
+			updateDisplayTableData.call(thisObject, this);
+			throw error;
+		}
+	}
+
+	protected async installCore() {
+		const installationCommands: Command[] = this.installationCommands;
+		if (!installationCommands || installationCommands.length === 0) {
+			throw new Error(`Cannot install tool:${this.displayName}::${this.description} as installation commands are unknown`);
+		}
+		for (let i: number = 0; i < installationCommands.length; i++) {
+			await this._platformService.runCommand(installationCommands[i].command,
+				{
+					workingDirectory: installationCommands[i].workingDirectory || this.downloadPath,
+					additionalEnvironmentVariables: installationCommands[i].additionalEnvironmentVariables
+				},
+				installationCommands[i].sudo,
+				installationCommands[i].comment,
+				installationCommands[i].ignoreError
+			);
+		}
+	}
+
+	protected async addInstallationSearchPathsToSystemPath(): Promise<void> {
+		const installationPath = await this.getInstallationPath();
+		const searchPaths = [installationPath, ...this.installationSearchPaths].filter(path => !!path);
+		this.logToOutputChannel(`Search Paths for tool '${this.displayName}': ${JSON.stringify(searchPaths, undefined, '\t')}`);
+		searchPaths.forEach(installationSearchPath => {
+			if (installationSearchPath) {
+				if (process.env.PATH) {
```

Comment:
This shouldn't be necessary - you're already filtering out undefined values on line 136

Reply:
Good Catch - I added the filter to remove undefined values later while this line was written earlier. Will remove.